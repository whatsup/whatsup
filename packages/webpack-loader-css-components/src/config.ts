import { cosmiconfig } from 'cosmiconfig'
import { ProcessOptions, PluginCreator, Plugin } from 'postcss'
import { LoaderContext } from './index'

type PluginsConfig = { [k: string]: boolean | PluginOptions } | (string | [string, PluginOptions])[]

type PluginOptions = {
    [k: string]: any
}

type Config = {
    plugins?: PluginsConfig
    processOptions?: ProcessOptions
}

const loadRawConfig = async (loaderContext: LoaderContext) => {
    const explorer = cosmiconfig('postcss')

    try {
        const found = await explorer.search()

        if (found === null || found.isEmpty) {
            throw 'Config is empty'
        }

        const { config, filepath } = found

        loaderContext.addBuildDependency(filepath)
        loaderContext.addDependency(filepath)

        if (typeof config === 'function') {
            const ctx = {
                env: loaderContext.mode,
                mode: loaderContext.mode,
                file: loaderContext.resourcePath,
                webpackLoaderContext: loaderContext,
                options: {},
            }

            return config(ctx) as Config
        }

        return config as Config
    } catch (e) {
        return {} as Config
    }
}

function* readPlugins(plugins: PluginsConfig): Generator<[string, PluginOptions], void, unknown> {
    if (Array.isArray(plugins)) {
        for (const item of plugins) {
            if (Array.isArray(item)) {
                const [plugin, options = {}] = item

                yield [plugin, options]
            } else {
                const plugin = item
                const options = {} as PluginOptions

                yield [plugin, options]
            }
        }
    } else {
        for (const [plugin, val] of Object.entries(plugins)) {
            if (!!val) {
                const options = typeof val === 'object' ? val : {}

                yield [plugin, options]
            }
        }
    }
}

const ignorePlugins = ['postcss-import', 'postcss-url', 'postcss-modules']

const preparePlugins = (plugins: PluginsConfig) => {
    const result = [] as Plugin[]

    for (const [plugin, options] of readPlugins(plugins)) {
        if (!ignorePlugins.includes(plugin)) {
            const module = require(plugin) as PluginCreator<PluginOptions>

            result.push(module(options) as Plugin)
        }
    }

    return result
}

const prepareProcessOptions = (options: ProcessOptions, loaderContext: LoaderContext) => {
    options.from = loaderContext.resourcePath
    options.to = undefined

    return options
}

export const loadConfig = async (loaderContext: LoaderContext) => {
    const raw = await loadRawConfig(loaderContext)
    const plugins = preparePlugins(raw.plugins || [])
    const processOptions = prepareProcessOptions(raw.processOptions || {}, loaderContext)

    return { plugins, processOptions }
}
