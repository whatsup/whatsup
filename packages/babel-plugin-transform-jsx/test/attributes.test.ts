import { transform } from './config/transform'

describe('transform attributes', function () {
    it('Should transform key prop', function () {
        expect(transform('<div key="Key" />')).toMatchSnapshot()
    })

    it('Should transform simple props', function () {
        expect(transform('<div prop1="value1" />')).toMatchSnapshot()
    })

    it('Should transform boolean prop', function () {
        expect(transform('<div focused />')).toMatchSnapshot()
    })

    it('Should transform props with expressions', function () {
        expect(transform('<div prop1={1 + 1} />')).toMatchSnapshot()
    })

    it('Should transform props spread', function () {
        expect(transform('<div {...props} />')).toMatchSnapshot()
    })

    it('Should transform different props variations', function () {
        expect(transform('<div key="Key" prop1="value1" prop2={2 + 2} {...props} />')).toMatchSnapshot()
    })

    it('Should transform different props with ref', function () {
        expect(transform('<div ref={myRef} key="Key" prop1="value1" prop2={2 + 2} {...props} />')).toMatchSnapshot()
    })

    it('Should throw error on namespaced attribute name', function () {
        expect(() => transform('<div ns:attr="val" />')).toThrowError('Namespaced attribute name is not supported')
    })
})
