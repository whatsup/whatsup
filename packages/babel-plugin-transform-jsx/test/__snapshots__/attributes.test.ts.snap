// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform attributes Should transform boolean prop 1`] = `
"import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, {
  focused: true
}, undefined);"
`;

exports[`transform attributes Should transform different props variations 1`] = `
"import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", \\"Key\\", {
  prop1: \\"value1\\",
  prop2: 2 + 2,
  ...props
}, undefined);"
`;

exports[`transform attributes Should transform key prop 1`] = `
"import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", \\"Key\\", undefined, undefined);"
`;

exports[`transform attributes Should transform props spread 1`] = `
"import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, { ...props
}, undefined);"
`;

exports[`transform attributes Should transform props with expressions 1`] = `
"import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, {
  prop1: 1 + 1
}, undefined);"
`;

exports[`transform attributes Should transform simple props 1`] = `
"import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, {
  prop1: \\"value1\\"
}, undefined);"
`;
